File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& s, vector<int>& p, long long target) {
       
        sort(p.begin(), p.end());
        
        int n = s.size();
        int m = p.size();
        
        vector<int> v1;
        
        for(int i=0; i<n; i++){
            
            // req number
            long long req = target/s[i] + (target%s[i] != 0);
            auto it = lower_bound(p.begin(), p.end(), req);
            
            if(it == p.end())     v1.push_back(0);
            else{
                int ind = it - p.begin();
                v1.push_back(m-ind);
            }
        }
        return v1;}
        
};

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        # spells.sort()
        potions.sort()
        
        # assuming len_spells<len_potions 
        n=len(potions)
        
        
        
        # binary search fn
        
        def binary_search(i):
            low,high=0,n-1
            while low<=high:
                mid=low+(high-low)//2
                if potions[mid]*i>=success:
                    if mid==0:
                        # print(i,n)
                        return n
                    elif potions[mid-1]*i>=success:
                        # print(mid)
                        high=mid-1
                    else:
                        # print(i,'i',mid)
                        return n-mid
                else:
                    low=mid+1
            return 0
                    
        
        
        
        possible_combns=[]
        for i in spells:
            possible_combns.append(binary_search(i))
        return possible_combns
        